File: src/test/java/com/github/kayjamlang/tests/expressions/NamedSingleExpressionTest.java
package com.github.kayjamlang.tests.expressions;

import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.KayJamLexer;
import com.github.kayjamlang.core.KayJamParser;
import com.github.kayjamlang.core.expressions.NamedExpression;
import com.github.kayjamlang.core.expressions.CallOrCreateExpression;
import com.github.kayjamlang.tests.TestsUtils;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertSame;

public class NamedSingleExpressionTest {

    @Test
    public void test() throws Exception {
        Expression expression = TestsUtils.parse("test -> println()");

        assertNotNull(expression);
        assertSame(NamedExpression.class, expression.getClass());

        NamedExpression namedContainer = (NamedExpression) expression;
        assertSame(CallOrCreateExpression.class, namedContainer.expression.getClass());
    }
}

File: src/test/scala/com/github/kayjamlang/tests/expressions/NamedSingleExpressionTest.scala
package com.github.kayjamlang.tests.expressions

import com.github.kayjamlang.core.containers.Container
import com.github.kayjamlang.core.{KayJamLexer, KayJamParser}
import com.github.kayjamlang.core.expressions.{CallOrCreateExpression, FunctionRefExpression, NamedExpression}
import org.junit.Assert.{assertNotNull, assertSame}
import org.junit.{BeforeClass, Test}

class NamedSingleExpressionTest {
  @Test
  @throws[Exception]
  def test1(): Unit = { // TODO: rename test!
    val parser = new KayJamParser(new KayJamLexer("test -> (x) { println(x) }"))

    val expression = parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[NamedExpression], expression getClass)

    val namedContainer = expression.asInstanceOf[NamedExpression]
    assertSame(classOf[FunctionRefExpression], namedContainer.expression getClass)

    val functionRefContainer = namedContainer.expression.asInstanceOf[FunctionRefExpression]
    assertSame(classOf[Container], functionRefContainer.expression getClass)

    val container = functionRefContainer.expression.asInstanceOf[Container]
    assertSame(classOf[CallOrCreateExpression], container.children.head getClass)
  }

  @Test
  @throws[Exception]
  def test2(): Unit = { // TODO: rename test!
    val parser = new KayJamParser(new KayJamLexer("test -> { println() }"))

    val expression = parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[NamedExpression], expression getClass)

    val namedContainer = expression.asInstanceOf[NamedExpression]
    assertSame(classOf[FunctionRefExpression], namedContainer.expression getClass)

    val functionRefContainer = namedContainer.expression.asInstanceOf[FunctionRefExpression]
    assertSame(classOf[Container], functionRefContainer.expression getClass)

    val container = functionRefContainer.expression.asInstanceOf[Container]
    assertSame(classOf[CallOrCreateExpression], container.children.head getClass)
  }

  @Test
  @throws[Exception]
  def test3(): Unit = { // TODO: rename test!
    val parser = new KayJamParser(new KayJamLexer("test -> println()"))

    val expression = parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[NamedExpression], expression getClass)

    val namedContainer = expression.asInstanceOf[NamedExpression]
    assertSame(classOf[FunctionRefExpression], namedContainer.expression getClass)

    val functionRefContainer = namedContainer.expression.asInstanceOf[FunctionRefExpression]
    assertSame(classOf[CallOrCreateExpression], functionRefContainer.expression getClass)
  }
}
